# Homework 1 Code and Report
AUFAR Rezka and LASBLEIS Alexandre 

### Step 1 : Simulate Rankings of Relevance for E and P (5 points)

In the first step you will generate pair s of rankings of relevance, for the production P and experimental E,
respectively, for a hypothetical query q . Assume a 3-graded relevance, i.e. {N, R, HR}. Construct all
possible P and E ranking pairs of length 5, for which E outperforms P.

Example:<br>
P: {N N N N N}<br>
E: {N N N N R}<br>
… <br>
P: {HR HR HR HR R}<br>
E: {HR HR HR HR HR}<br>

(Note 1: If you do not have enough computational power, sample 1000 pair uniformly at random to show
your work.)
##### Code

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import random as rand

In [3]:
#convert int to relevance
#_list : list of integer to be converted
#return : list of string where every integer has been replaced by its relevance interpretation
def to_string(_list): 
    L=[]
    for i in _list:
        if i==0:
            L.append("N ")
        elif i==1:
            L.append("R ")
        elif i==2:
            L.append("HR")
        else: 
            L.append("? ")
    return L

In [4]:
pairs = [] #variable used to store all pairs 
Ps = []
Es = []

#Create all combinaison for P, there are 3^5 = 243 possibilities since P is of length 5 and each member have 3 possible values:
# 0 (representing "N" ), 1 (representing "R), 2 (representing "HR)
for p_possibility in range(243): 
    P=[None]*5
    #create all five values for P 
    for p_pos in range(5):
        P[p_pos]= p_possibility // (np.power(3,p_pos)) %3 #formula to create all combinaison
    #we do the same for E    
    for e_possibility in range(243): 
        E=[None]*5
        for e_pos in range(5):
            E[e_pos]= e_possibility // (np.power(3,e_pos)) %3 #formula to create all combinaison
        pairs.append((P,E)) #adding the pair since every pair is valid
        #Memorize all created Es but only once
        if p_possibility==0:
            Es.append(E)
        #memorize all created Ps
    Ps.append(P)

In [48]:
#TESTING
print(len(pairs)) #= 59049 which is equal to the expected length = 3 ^10 = 59049
#print(pairs) #print all pairs but this takes a lot of place

#print a random pair to show that it makes sense 
r = rand.randint(0,59049-1)
print("P: " , to_string(pairs[r][0]), "E: ",to_string(pairs[r][1]))

#check if Es is as expected 
print(len(Es))#expect 243 = 3^5 
print(Es)

59049
('P: ', ['HR', 'HR', 'N ', 'R ', 'N '], 'E: ', ['R ', 'HR', 'HR', 'R ', 'N '])
243
[[0, 0, 0, 0, 0], [1, 0, 0, 0, 0], [2, 0, 0, 0, 0], [0, 1, 0, 0, 0], [1, 1, 0, 0, 0], [2, 1, 0, 0, 0], [0, 2, 0, 0, 0], [1, 2, 0, 0, 0], [2, 2, 0, 0, 0], [0, 0, 1, 0, 0], [1, 0, 1, 0, 0], [2, 0, 1, 0, 0], [0, 1, 1, 0, 0], [1, 1, 1, 0, 0], [2, 1, 1, 0, 0], [0, 2, 1, 0, 0], [1, 2, 1, 0, 0], [2, 2, 1, 0, 0], [0, 0, 2, 0, 0], [1, 0, 2, 0, 0], [2, 0, 2, 0, 0], [0, 1, 2, 0, 0], [1, 1, 2, 0, 0], [2, 1, 2, 0, 0], [0, 2, 2, 0, 0], [1, 2, 2, 0, 0], [2, 2, 2, 0, 0], [0, 0, 0, 1, 0], [1, 0, 0, 1, 0], [2, 0, 0, 1, 0], [0, 1, 0, 1, 0], [1, 1, 0, 1, 0], [2, 1, 0, 1, 0], [0, 2, 0, 1, 0], [1, 2, 0, 1, 0], [2, 2, 0, 1, 0], [0, 0, 1, 1, 0], [1, 0, 1, 1, 0], [2, 0, 1, 1, 0], [0, 1, 1, 1, 0], [1, 1, 1, 1, 0], [2, 1, 1, 1, 0], [0, 2, 1, 1, 0], [1, 2, 1, 1, 0], [2, 2, 1, 1, 0], [0, 0, 2, 1, 0], [1, 0, 2, 1, 0], [2, 0, 2, 1, 0], [0, 1, 2, 1, 0], [1, 1, 2, 1, 0], [2, 1, 2, 1, 0], [0, 2, 2, 1, 0], [1, 2, 2, 1, 0], [2, 2, 2,

##### Explanations
In order to contruct all possible pairs of P and E with a 3-graded relevance {N, R, HR}, we first encode those relevance as integers: <br>
* 0 for N <br>
* 1 for R <br>
* 2 for HR <br>
This encoding allows us to do easy comparison between elements.
Then we created all pairs using this formula : $value = \frac{possibility}{3^{pos}} [3]$ with $possibility \in \{0,...,243\}$ and $ pos \in \{0,1,2,3,4\}$.

This give the following sequence depending on the position of the element in the list: <br>
* 0: $\{0,1,2,0,1,2,0,1,2,0,1,2...\}$
* 1: $\{0,0,0,1,1,1,2,2,2,0,0,0...\}$
* 2: $\{0,0,0,0,0,0,0,0,0,1,1,1,...\}$
* ... and so on

This formula is then used for the production P and the experiment E. 

##### Analysis
According to the formula we showed before, we should find all the possible combination. Futhermore, the number of pairs found is equal to the one we expected $3^{10}$ and the number of E and P are $3^{5}$ each which is also what is expected. 


### Step 2 : Implement Evaluation Measures (15 points)

Implement 1 binary and 2 multi-graded evaluation measures out of the 7 measures mentioned above.

(Note 2: Some of the aforementioned measures require the total number of relevant and highly relevant
documents in the entire collection – pay extra attention on how to find this)

TODO: ================================================================

========================================================================

===========================================================================

add if security if the number of relevant is null ! 

TODO: Remove for evaluation 



#Recall at rank k 
def Binary_measure_recal(test_set, r):
    num_test_rel = 0 
    total_num_rel=0
    for i in range(len(test_set)):
        if test_set[i]>0:
            if i<r:
                num_test_rel+=1
            total_num_rel+=1
    return float(num_test_rel) / total_num_rel

#Average precision at rank k 
def Binary_measure_avg_precision(test_set, r):
    num_rel=0
    sum_precision = 0 
    total_num_rel=0
    temp="("
    for i in range(len(test_set)):
        if test_set[i]>0:
            if i<r:
                num_rel+=1
                temp+=" "+str(num_rel)+"/"+str(i+1)+"+"
                sum_precision+=float(num_rel)/(i+1)
            total_num_rel+=1
    temp=temp[:len(temp)-1]+")"
    #print(temp,"/",total_num_rel)
    return float(sum_precision) / total_num_rel
#Normalized discount cumulative gain nDCG at rank r
def nDCG(test_set, r):
    orderedSet= list(test_set)
    orderedSet.sort(reverse=True)
    return DCG(test_set, r) / DCG(orderedSet, r)

In [20]:
#BINARY 
#Precision at rank r 
#test_set: the collection to be tested 
#r: the rank for which we test the collection 
#return : the precision 
def Binary_measure_precision(test_set, r):
    assert r <= len(test_set)
    assert r!= 0
    num_test_rel = 0 
    for i in range(r):
        #According to our notation, if test_set[i]>0, the document is relevant
        if test_set[i]>0:
            num_test_rel+=1
    return float(num_test_rel) / r #precision is number of relevant doc/ number of doc considered

#MULTI GRADED 
#Discount cumulative gain DCG at rank r
#test_set: the collection to be tested 
#r: the rank for which we test the collection 
#return : the gain 
def DCG(test_set, r):
    assert r <= len(test_set)
    assert r!= 0
    res=0
    for k in range(1,r+1):#r is the rank so it starts at 1
        rel_r = test_set[k-1]#index of element = rank -1 since rank starts at 1 and index at 0
        res+=float(np.power(2,rel_r)-1)/(np.log2(1+k))
    return res

#Rank Biased Precision w ith persistence parameter 𝜃 =0.8
#test_set: the collection to be tested 
#theta: the theta parameter of the RBP formula
#return : the precision
def RBP (test_set,theta=0.8):
    res=0
    for k in range(1,len(test_set)+1):
        rel_k=test_set[k-1]
        res+= rel_k * np.power(theta,k-1) * (1.0-theta)
    return res
        

In [25]:
test= [1,0,1,1,0,1,0,0,0,1,0,1,1,0,1,0,0,1,0,1]#length 20, inspired of the example p 25 in Lecture 1b
test_2 = [0,0,0,0,0]
r=11
print('Precision',Binary_measure_precision(test,r))
print('DCG', DCG(test,r))
print('RBP', RBP(test))

('Precision', 0.45454545454545453)
('DCG', 2.5759485714993033)
('RBP', 0.56988530654229552)


##### Explanations
For Precision and DCG, we need to ensure that rank k does not exceed the documents length, and k cannot be zero.
##### Analysis
The result for Precision is highly dependant on the number of non-relevant in the cut-off collection. As k increases, the more non-relevant exists in the cut-off, the lower the precision will be. <br>
For DCG, the result depends on the degree of the relevance and the order of it. If the highest relevant score exists in the first position of the cut-off, the higher DCG will be. <br>
For RBP, we disregard rank and therefore we have the same result no matter the rank is.

### Step 3 : Calculate the measure (5 points)

For the three measures and all P and E ranking pairs constructed above calculate the difference: 𝛥measure
= measure E -measure P . Consider only those pairs for which E outperforms P.

In [52]:
def outperforms(methond_name,E,P,rank):
    if methond_name=="precision":
        delta = Binary_measure_precision(E,rank)-Binary_measure_precision(P,rank)
        return (delta>0), delta 
    elif methond_name=="DCG":
        delta=DCG(E,rank)-DCG(P,rank)
        return (delta>0), delta 
    elif methond_name=="RBP":
        delta=RBP(E)-RBP(P)
        return (delta>0), delta 
        
#Measures chosen : precision, DCG, RBP 
outperf_pairs_prec=[]
pairs_set_prec = []
d_measure_prec=[]

outperf_pairs_DCG=[]
pairs_set_DCG = []
d_measure_DCG=[]

outperf_pairs_RBP=[]
pairs_set_RBP = []
d_measure_RBP=[]

rank = 4
for i in range(len(pairs)):#pairs made of (P,E)
    outperforms_prec,delta_prec= outperforms("precision",pairs[i][1],pairs[i][0],rank)
    outperforms_DCG,delta_DCG= outperforms("DCG",pairs[i][1],pairs[i][0],rank)
    outperforms_RBP,delta_RBP= outperforms("RBP",pairs[i][1],pairs[i][0],rank)

    if(outperforms_prec):
        outperf_pairs_prec.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_prec.append(''.join(str(x) for x in temp))
        d_measure_prec.append(delta_prec)
        
    if(outperforms_DCG):
        outperf_pairs_DCG.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_DCG.append(''.join(str(x) for x in temp))
        d_measure_DCG.append(delta_DCG)
        
    if(outperforms_RBP):
        outperf_pairs_RBP.append(pairs[i])
        temp = pairs[i][1] + pairs[i][0]
        pairs_set_RBP.append(''.join(str(x) for x in temp))
        d_measure_RBP.append(delta_RBP)     


In [69]:
#TESTING
length = min(100,len(outperf_pairs_prec))
#print len(outperf_pairs_prec)
#print len(outperf_pairs_DCG)
#print len(outperf_pairs_RBP)

print len(set.intersection(set(pairs_set_prec), set(pairs_set_DCG), set(pairs_set_RBP)))
print rand.sample(set.intersection(set(pairs_set_prec), set(pairs_set_DCG)),1)

# pick result that is in prec but is not in DCG
print rand.sample(set(pairs_set_prec) - set(pairs_set_DCG),1)
# pick result that is in DCG but is not in prec
print rand.sample(set(pairs_set_prec) - set(pairs_set_DCG),1)

#for j in range(length):
#    print("P:",to_string(outperf_pairs_prec[j][0]),"E:",to_string(outperf_pairs_prec[j][1]))
#    print(d_measure_prec[j])

17184
['1102100120']
['1012222002']
['0221222001']


##### Explanations
For precision, DCG, and RBP, we calculate the difference between all combination of E and P that we have generated in Step 1, then we append every difference that is greater than zero, which means that E outperforms P.
##### Analysis
For each three measures, we have different pairs in which E outperforms P. There are 17184 pairs which E outperforms P in all three measures have. However, there exists some pairs which the measures does not yield the same answer. <br> 
For example, E=[0,1,1,1,0] outperforms P=[1,0,2,0,0] in precision, but not in DCG. E=[1,2,0,0,1] outperforms P=[2,0,0,0,2] in DCG, but not in precision. <br>
This indicates that in real world application, it is important to implement different kinds of measure so we will have broader feedback to optimize the algorithm.